File: python-transcription-service/transcribe/app/Services/Transcribe/Clients/WhisperClient.php
<?php

namespace App\Services\Transcribe\Clients;

use App\Services\Transcribe\TranscribeClientInterface;
use App\Services\Transcribe\TranscribeServiceException;
use GuzzleHttp\Client;
use GuzzleHttp\Exception\GuzzleException;
use GuzzleHttp\RequestOptions;

class WhisperClient implements TranscribeClientInterface
{
    public function __construct(private readonly Client $client)
    {
    }

    /**
     * @inheritdoc
     */
    public function process($f_handle, int $channel, string $locale): array
    {
        $locale = explode('_', str_replace('-', '_', $locale))[0];

        $multipart = [
            [
                'name'     => 'model',
                'contents' => 'istupakov/parakeet-tdt-0.6b-v3-onnx',
            ],
            [
                'name'     => 'response_format',
                'contents' => 'verbose_json',
            ],
            [
                'name'     => 'pause_threshold',
                'contents' => 0.5,
            ],
            [
                'name'     => 'file',
                'contents' => $f_handle,
            ],
        ];

        $query = [
//            'response_format' => 'json',
        ];

        if ($locale) {
            $query['language'] = $locale;
        }

        $start_time = time();
        try {
            $response = $this->client->post('', [
                RequestOptions::MULTIPART => $multipart,
                //                RequestOptions::QUERY     => $query
            ]);
        }
        catch (GuzzleException $e) {
            throw new TranscribeServiceException($e->getMessage(), $e->getCode(), $e);
        }

        $duration_process = time() - $start_time;
        $result = $response->getBody()->getContents();

        $result = ((array)json_decode($result, true))['segments'] ?? [];

        $items = [];
        foreach ($result as $item) {
            $items[] = [
                'channel'    => $channel,
                'start_time' => round($item['start'], 2),
                'end_time'   => round($item['end'], 2),
                'text'       => mb_trim($item['text']),
            ];
        }

        return $items;
    }
}

File: python-microservice-repo/py-services/services/transcribe/whisper_client.py
#!/usr/bin/env python3
"""
Клієнт до Whisper API для транскрипції аудіо.
"""
import httpx

from shared.logger import get_logger

logger = get_logger("transcribe")


class WhisperClient:
    """HTTP клієнт до Parakeet ASR API (Макс)."""

    def __init__(self, url: str, model: str, connect_timeout: float = 10, timeout: float = 800):
        self.url = url.rstrip("/") + "/parakeet_asr"
        self.model = model
        self.timeout = httpx.Timeout(timeout, connect=connect_timeout)

    async def transcribe(self, audio_data: bytes, filename: str, locale: str) -> list[dict]:
        """Відправляє аудіо на Parakeet ASR API, повертає список сегментів.

        Args:
            audio_data: Бінарні дані аудіофайлу.
            filename: Ім'я файлу (потрібне для multipart).
            locale: Мова (наприклад 'uk', 'ru', 'en').

        Returns:
            Список сегментів:
            [{"channel": 1, "start_time": 1.23, "end_time": 2.45, "text": "привіт"}, ...]
        """
        lang = locale.split("_")[0].split("-")[0]

        files = {"file": (filename, audio_data, "audio/mpeg")}
        data = {
            "model": self.model,
            "response_format": "verbose_json",
            "pause_threshold": "1.5",
        }
        if lang:
            data["language"] = lang

        async with httpx.AsyncClient(timeout=self.timeout) as client:
            response = await client.post(self.url, files=files, data=data)
            response.raise_for_status()

        result = response.json()
        segments = result.get("segments", [])

        items = []
        for seg in segments:
            text = seg.get("text", "").strip()
            if not text:
                continue
            items.append({
                "channel": 1,
                "start_time": round(float(seg.get("start", 0)), 2),
                "end_time": round(float(seg.get("end", 0)), 2),
                "text": text,
            })

        return items
